Bootstrap protocol implementation

busykai opened the pull request:
Implemented parts:

1. Initial connection;
2. Application reload;
3. Only one `APP_HOST` per instance;
4. Only one `SIM_HOST` per instance;
5. `APP_HOST` reloads on `SIM_HOST` reload.

msftclas commented:
Hi __@busykai__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

albertinad commented on src/app-host/app-host.js:
```diff
@@ -57,38 +100,6 @@ function getCordova() {
     return cordova;
 }
 
```
Plugins simulation (app-host side) is loaded and initialized without being attached to the overall initialization. That could cause issues like requesting exec calls to sim-host side when sim-host is `ready` but plugins are not ready yet because they are initializing. So this exec calls in sim-host fails, because there is no handler, when actually it is.

busykai replied:
Thanks for catching this! It was behaving weirdly on me, but I didn't get to actually root-cause it.

albertinad commented on src/app-host/app-host.js:
```diff
@@ -5,41 +5,85 @@ var Messages = require('messages');
+            require('xhr-proxy').init(); 
+        });
+
+        channel.onNativeReady.fire();
```
Since plugin's simulation code acts as the native side of the plugin, those should be initialized and ready before firing `onNativeRead`, what do you think @busykai ?

busykai replied:
Yes, `init` message is sent to `APP_HOST` only after `SIM_HOST` tells the server it's ready.

albertinad commented on src/server/socket.js:
```diff
@@ -1,103 +1,204 @@
-            // Set up telemetry if necessary.
-            if (config.telemetry) {
-                socket.emit('init-telemetry');
+    socket.once('start', handleStart);
```
This function is called even when there's no sim-host instance connected. Maybe this scenario is not taken into account yet. To reproduce it, launch the simulation, you will get sim-host and app-host connected, then close sim-host so it gets disconnected, and close app-host. Load app-host one more time, and you will si an exception in the server log: `TypeError: Cannot read property 'once' of undefined`.

busykai replied:
thank you for this catch!

albertinad commented on src/server/socket.js:
```diff
@@ -1,103 +1,204 @@
+            log.log(type + ' disconnected to the server');
+            switch (type) {
+                case APP_HOST:
+                    resetAppHostState();
```
When app-host is disconnected we reset the state, that means creating a new promise and attach a callback for when the promise is resolved, but when the app-host sends the `register-app-host` message, the promise is resolved. And the handler gets executed assuming that there's a connected sim-host instance.

busykai replied:
@albertinad, is this comment related to the crash on when `SIM_HOST` gets connected and then disconnected? I cannot figure out if this is something I addressed already or it's still pending. :)

albertinad replied:
@busykai yes! that is related to that issue. I will try your new changes ASAP :)

albertinad commented on src/server/socket.js:
```diff
@@ -1,103 +1,204 @@
     livereload = require('./live-reload/live-reload-server'),
-    telemetry = require('./telemetry-helper');
+    telemetry = require('./telemetry-helper'),
+    q = require('q');
```
The variable name should be with uppercase, for consistency with other usages and since `Q` is the global when no using amd loader.

albertinad commented on src/app-host/app-host.js:
```diff
@@ -5,93 +5,184 @@ var Messages = require('messages');
+            default:
+                channel.onNativeReady.fire();
+                break;
+        }
```
Nice!!

albertinad commented on src/app-host/app-host.js:
```diff
@@ -5,93 +5,184 @@ var Messages = require('messages');
-socket.on('init-xhr-proxy', function (data) {
-    require('xhr-proxy').init();
-});
+        channel.onNativeReady.fire();
```
This is duplicated. I guess that you want to fire `onNativeReady` as you're doing below.

busykai replied:
Thanks! That must be rebase artifact.

albertinad commented:
@busykai, When app-host is reload and about to be sent, sometimes sim-host is requested to be reloaded too. So a `refresh` message is sent to sim-host, and immediately a [function to invalidate sim-host](https://github.com/Microsoft/cordova-simulate/blob/master/src/server/sim-files.js#L43) instance is made. Calling `simSocket.invalidateSimHost();` sets to undefined to the sim-host socket reference that we keep. I think that this is not needed. When sending `refresh`, sim-socket will reload itself, so we are going to receive a disconnection and then, a new sim-host connection.
That call, sometimes causes an error, as you can see in this section of the log:
```
SIM:  Creating app-host.js
SIM:  Emitting 'refresh' to sim-host (pending connection)
Uncaught top-level Node.js exception: TypeError: Cannot read property 'push' of undefined
TypeError: Cannot read property 'push' of undefined
    at Object.emitToHost (<path>/cordova-simulate/src/server/socket.js:268:27)
    at Object.validateSimHostPlugins (<path>/cordova-simulate/src/server/sim-files.js:42:19)
    at <path>/cordova-simulate/src/server/server.js:52:18
From previous event:
    at streamAppHostHtml (<path>/cordova-simulate/src/server/server.js:50:8)
    at Layer.handle [as handle_request] (<path>/components/server/emulator/cordova-simulate/xdk-cordova-serve/node_modules/express/lib/router/layer.js:95:5)
```

busykai commented:
1. I'll refactor the server code to avoid crashes and print an error instead. It is critical for Intel XDK usecase that server is stable.
2. We no longer need pending emits during the bootstrap as the protocol ensures there must be none.
3. I'll also implement what's left of the protocol spec (except for multiple app-hosts).

albertinad commented:
That sounds good! thank you @busykai :)

albertinad commented on src/app-host/app-host.js:
```diff
@@ -5,93 +5,187 @@ var Messages = require('messages');
+
+        socket.on('exec-failure', function (data) {
+            console.log('exec-failure:');
+            console.log(data);
```
We can change this to
```
console.log('exec-failure:', data);
```

busykai replied:
I think we should remove it altogether. I just moved the code around. It's not a minor thing, we should be careful to avoid contaminating application console.

What I think we should add is a warning for `exec()` calls that did not return in a reasonable period of time (5 sec?).

albertinad replied:
That is better! I'm not so sure about the warning in the reasonable period of time. How does cordova handles that? Does the call to native side have a maximum "waiting" time?

busykai replied:
I don't think cordova keeps track of that (among the implementation for the platforms I've seen). We could though, but it will not be part of this PR. We should have a "verbose" mode for `APP_HOST` where things such as exec traces and unreturned calls are printed.

albertinad commented on src/app-host/app-host.js:
```diff
@@ -5,93 +5,187 @@ var Messages = require('messages');
+            var execCacheInfo = execCache[data.index];
+            if (execCacheInfo.fail) {
+                execCacheInfo.fail(data.error);
+            }
```
Does it makes sense to check if `execCacheInfo.fail` is a function, before executing it instead of the current validation?
```
if (typeof execCacheInfo.fail === 'function') {
    execCacheInfo.fail(data.error);
}
```
Also, could we have an scenario in which execCache[data.index] is undefined? If exists such scenario, maybe we can add a better validation too.
I remember that a while ago, there were situations in which execCacheInfo was undefined (in sim-host side), but it was fixed in one of the merges by sending the register event at last.

busykai replied:
I think it's a good idea to check `execCacheInfo` is actually there. Type check seems to be excessive for me: it would be a plugin simulation bug. We need to simplify a task of adding simulation support for a plugin author, it would be helpful there. I'm thinking of whether it needs to be done now or later (as part of the "developing simulation support" feature later, for example).

Anyways, I agree... I think this should be the condition for the happy scenario:
```javascript
if (execCacheInfo && typeof execCacheInfo.fail === 'function') {
```

albertinad commented on src/server/socket.js:
```diff
@@ -1,115 +1,194 @@
+
+    var socket = hostSockets.SIM_HOST;
+
+    socket.on('exec-success', function (data) {
```
Maybe you can use `suscribeTo` function that internally does this for you.

busykai replied:
Yes, it was the idea. I will clean this up with the final review.

albertinad commented:
@busykai This looks really great! I've been playing a while and it works. If we find some corner cases or some scenarios that we've missed we could fix it later, but so far, I didn't find anything.

albertinad commented on src/server/socket.js:
```diff
@@ -1,115 +1,193 @@
+    // Set up live reload if necessary.
+    if (config.liveReload) {
+        log.log('Starting live reload.');
+        livereload.init(socket);
```
This should be updated too, socket is not available.

albertinad commented:
@busykai I have one more comment about these changes.
There's a feature that enables simulation code, in both hosts, to emit and listen for events, and also to define methods and execute those methods. There are cases in which the sim-host side of a simulation plugin, needs to get some information that is only available on the app-host side. For example, the device simulation, we need to get the current cordova version, and to do that, we `register` a method in app-host side, that is called by the sim-host side to show the cordova version in the widget.
There's a problem right now: we never get the value, because we're a executing the method before it can be registered or when cordova is not attached to the window yet. With your changes I can propose a fix, but it is not a good one: register a method and also emit an event, and in the sim-host side, call that method, and listen for the event. That is a workaround, but we still need a better solution, and common solution that address all plugin's issues related to this, when they register and execute methods.
Since now we have more clear the steps of the initialization, we know when sim-host is ready, and also when the simulation modules are ready on app-host, so we can queue the messages `plugin-method` (and maybe `plugin-message`?) until both sim-host and app-host are ready, and start to emit them.
What do you think?

dlebu commented on src/app-host/app-host.js:
```diff
@@ -5,93 +5,184 @@ var Messages = require('messages');
+            case 'windows':
+            /* falls through */
+            default:
+                channel.onNativeReady.fire();
```
Does `android` go to the default case?

busykai replied:
Yes! Do you think we should explicitly name it here? It might be worth it.

dlebu replied:
I think that's a good idea.

guillaumejenkins replied:
Since `windows` is already there with a `/* falls through */` comment, I think it's a good idea to add `android` as well

dlebu commented on src/server/socket.js:
```diff
@@ -1,115 +1,193 @@
+        emitTo(DEBUG_HOST, data.message, data.data);
+    });
+
+    subscribeTo(SIM_HOST, 'telemetry', function (data) {
```
Would it make sense to move this `subscribeTo` call in the next `if` statement, such that we subscribe only if telemetry is enabled?

dlebu replied:
I see that telemetry is sent over the socket even if not enabled, so please ignore my comment if it does not make sense.

guillaumejenkins replied:
Telemetry should not be sent over the socket if not enabled. The `init-telemetry` message below is needed by the client side for telemetry to start being sent. For that reason, I think it does make sense to move the subscribe inside the `if`

dlebu commented on src/server/socket.js:
```diff
@@ -1,115 +1,193 @@
+            log.log('APP_HOST connected to the server');
+            if (hostSockets[APP_HOST]) {
+                log.log('Overriding previously connected APP_HOST');
+                resetAppHostState();
```
For the current implementation, only 1 APP is supported, right?

busykai replied:
Yes. I believe I implemented the override here because while testing the refresh/reload scenario I found that we couldn't not rely on old APP_HOST `disconnect` getting before the new APP_HOST `register` message. Perhaps, it could be resolved by closing the socket or explicitly unregistering in `onbeforeunload`, but the new-registration-overrides type of policy is the safest.

dlebu commented on src/server/socket.js:
```diff
@@ -119,23 +197,48 @@ function init(server) {
+            }
+            hostSockets[type] = undefined;
+            log.log(type + ' disconnected to the server');
+            switch (type) {
```
Do we need to handle DEBUG_HOST in a similar fashion here? @guillaumejenkins

busykai replied:
I moved the `DEBUG_HOST` disconnect handler here.

dlebu commented on src/sim-host/protocol/socket.js:
```diff
@@ -11,52 +11,68 @@ module.exports.initialize = function (pluginHandlers, serviceToPluginMap) {
     });
+
+    socket.on('app-plugin-list', function (data) {
+        // TODO: process the list of plugins
```
Please open an issue so we don't forget about this `TODO` if it is still needed.

dlebu commented:
The change looks good. I had a couple of comments regarding telemetry and the debug host. I know those features came in after you started this PR, thus I am trying to open the conversation on improving those parts of the code as well.

guillaumejenkins commented on src/sim-host/protocol/socket.js:
```diff
@@ -11,52 +11,68 @@ module.exports.initialize = function (pluginHandlers, serviceToPluginMap) {
+                handler(success, failure, data.args);
+            }
+
+            telemetry.sendClientTelemetry('exec', telemetryProps);
```
The current implementation of the telemetry-helper under should already take care of that. It will never send telemetry unless it has been told to initialize itself. So, as long as we only call `telemetryHelper.init()` when the server asks it, these calls will not be sent to the server needlessly.

dlebu replied:
Allright, the PR is not up to date with the latest changes from master. Sounds good. Thanks for confirming.

albertinad commented on src/sim-host/protocol/socket.js:
```diff
@@ -8,77 +8,93 @@ var registerOnInitialize = false;
+                }
+
+                var success = data.hasSuccess ? getSuccess(index) : null;
+            var failure = data.hasFail ? getFailure(index) : null;
```
Fix indentation here.

busykai replied:
I need to fix linting to include some formatting checks... Half of the file has been indented wrongly.

albertinad commented:
This looks good to me! I believe that we are really good to merge it to master, and if we find areas of improvements, we can do it from there. We've been using this new protocol for a while, and all major issues have been addressed and it works really great!

dlebu commented:
LGTM! :shipit:

guillaumejenkins commented on src/app-host/app-host.js:
```diff
@@ -7,10 +7,61 @@ var telemetry = require('telemetry-helper');
+    });
+}
 
 function setCordova(originalCordova) {
```
Should we rename the function now that it does pretty much all the app-host initialization? The name suggests it only overrides the existing cordova, when in reality it does several extra things (socket handlers for exec calls, telemetry, XHR, touch events and live reload, initializing plugins, etc)... Or, maybe refactor all the non-cordova stuff to a new function? That way we would have a clear separation of cordova initialization and app-host-specific initialization... Something like `initCordova()` and `initAppHost()`?

guillaumejenkins commented:
The changes look great!

I'm still having a slightly hard time wrapping my head around one thing though: with these changes, what triggers the whole app-host initialization (set up socket handlers, initialize plugins, initialize telemetry, etc) is setting `window.cordova = ...`. I find it a bit unintuitive that all the initialization is done by setting a property on the `window` global object.

Now, I couldn't find anywhere in our code a call to `window.cordova = ...`, so I have to assume it comes from `cordova-serve`, and using this setter is a way to "listen" for that event. If that's the case, I think it might be clearer to let all the initialization logic run at the "root" of app-host.js (meaning it will be run when the script is loaded), and have a promise that is resolved in the `window.cordova` setter, and we can then defer the `socket.emit('register-app-host'` call until both the initialization AND the `window.cordova` patching has been done.

It might be confusing; here's what I mean:

```
// app-host.js
var livereload = require('./live-reload-client');
// ... Other require statements and code

var cordovaPatchedDeferred = Q.defer();

function initialize() {
    socket.on('start-live-reload', function () {
        livereload.start(socket);
    });
    // ... All the other code to set up socket handlers, initialize plugins, etc... basically anything not related to window.cordova patching

    return Q.resolve();
}

function setCordova(originalCordova) {
    cordova = originalCordova;
    cordova.define.remove('cordova/exec');
    // ... Other logic to patch Cordova (but nothing related to app-host specific things, like socket handlers, plugin clobbers, etc)

    cordovaPatchedDeferred.resolve();
}

// The following gets executed when the script gets loaded
Q.all([cordovaPatchedDeferred.promise, initialize()]).then(function () {
    // window.cordova has been set and app-host initialization has been done
    socket.emit('register-app-host');
};
```

As you can see, app-host is now responsible for its own initialization, but still waits for `cordova-serve` to define `window.cordova` before doing things related to cordova patching. I find this separation makes it much easier to understand the life cycle of app-host. I think it also takes care of a race condition: in your proposed changes, what happens if the server emits `init` to app-host before `cordova-serve` does `window.cordova = ...`? In that case, `socket.on('init', ...)` will not be registered, and so app-host might miss the initialization message.

Anyway, sorry for this huge wall of text, I just wanted to share my opinion. I am totally open to discussion about all of this, so everyone, please let me know what you think! It doesn't really change the approach / logic; what I propose is merely a refactor to separate the 2 different type of initialization that we need in app-host: things related to cordova patching, and things related to socket / plugin handlers.

busykai commented:
@guillaumejenkins, this could seem a bit counter-intuitive, I agree. And it potenitally could be improved (in sense of code readability). However, we inevitably have to "hack" cordova because it does not have mechanisms in place which would a better designed/structured approach to simulation initialization. `app-host.js` is injected at the very beginning of the head and is therefore executed first which makes patching and the rest of it possible (since it is loaded and exoverride the setter for a global).

Parallel initialization cannot be done since the protocol requires a `cordova` instance to be present already (it references it). Note that neither `cordova` patching nor socket handlers setup are a deferred action, so use of promises is not needed. Setting `cordova` is an action in the future for which we could use promise.

So we could refactor to do something like:

```javascript
function init() {
    // setup protocol socket listeners
    socket.emit('register-app-host');
}

function setCordova(cordova) {
    // patch cordova
    // create sim channel and delay deviceReady
   init();
}
```
I was thinking about something like this. However, it will be even more confusing to me. IMO it's incorrect to think that protocol setup and patching `cordova` are two different and independent activities. Also, we need to `init()` to execute synchronously (since it subscribes to some `cordova` events). 

About this specific point:
> I think it also takes care of a race condition: in your proposed changes, what happens if the server emits init to app-host before cordova-serve does `window.cordova = ...`?

Server takes care of all the timing-sensitive messages. `init` will never be sent before `windows.cordova = ...` since `register-app-host` is only send upon assigning an object to `window.cordova`. Event sequencing is quite fine-tuned by now.

guillaumejenkins commented:
Thank you very much for the clarifications! I was not 100% clear on how everything gets initialized and I now see that leaving things this way is necessary.

In that case, do you think it would be possible to rename `setCordova` to `initializeAppHost` or something similar? Just that simple change would go a long way towards indicating that the current `setCordova` method is actually the "entry point" of app-host. Even if it's stilled called from the `window.cordova` setter, it would make things slightly more intuitive, at least to me. What do you think?

busykai commented:
@guillaumejenkins, you're welcome! i hope they make sense. i've renamed the function to `setCordovaAndInitialize`. are you OK with the name?

I'll test it a little bit more over the weekend then it will be ready to merge.

guillaumejenkins commented:
@busykai Yes, I think that's clearer. Basically I just wanted a clearer way to find the entry point in app-host.js, and I think the new name helps.

Thanks, and consider me signed off!

busykai commented:
@guillaumejenkins, @dlebu, @albertinad I've addressed all the comments except for timing of `plugin-*` messages (which I'm working on right now). I'll merge this PR by tomorrow.

busykai commented:
I'm going to merge it as it as and file a separate issue to address plugins' message timing.

dlebu commented:
Sounds good!

busykai commented:
Filed #90. (BTW, allow me to take it.) Merging this one.

albertinad commented:
Awesome!!
